clarity/contracts/arkadiko-auction-engine-v4-1.clar

```diff
@@ -0,0 +1,440 @@
+    (asserts! (is-eq (contract-of liquidation-pool) (unwrap-panic (contract-call? .arkadiko-dao get-qualified-name-by-name "liquidation-pool"))) (err ERR-NOT-AUTHORIZED))
+    (asserts! (is-eq (contract-of liquidation-rewards) (unwrap-panic (contract-call? .arkadiko-dao get-qualified-name-by-name "liquidation-rewards"))) (err ERR-NOT-AUTHORIZED))
+    (asserts!
+      (or
+        (is-eq (contract-of reserve) (unwrap-panic (contract-call? .arkadiko-dao get-qualified-name-by-name "stx-reserve")))
+        (is-eq (contract-of reserve) (unwrap-panic (contract-call? .arkadiko-dao get-qualified-name-by-name "sip10-reserve")))
+      )
+      (err ERR-NOT-AUTHORIZED)
+    )
+    (asserts! 
+      (or
+        (and token-is-stx (is-eq (contract-of ft) .xstx-token))
+        (and (not token-is-stx) (is-eq (contract-of ft) token-address))
+      ) 
+      (err ERR-TOKEN-TYPE-MISMATCH)
+    )
+
+    (if (is-eq usda-to-use u0)
+      true
+      (begin
+        ;; Get USDA from fund
+        (try! (as-contract (contract-call? liquidation-pool withdraw usda-to-use)))
+
+        ;; Burn USDA, get collateral token
+        (try! (as-contract (contract-call? .arkadiko-dao burn-token .usda-token usda-to-use (as-contract tx-sender))))
+        (try! (as-contract (contract-call? vault-manager redeem-auction-collateral ft token-string reserve collateral-sold (as-contract tx-sender)))) 
+
+        ;; Deposit collateral token
+        (try! (as-contract (contract-call? liquidation-rewards add-reward block-height token-is-stx ft collateral-without-fee)))
+
+        ;; Update auction
+        (map-set auctions { id: auction-id } (merge auction {
+          total-collateral-sold: (+ (get total-collateral-sold auction) collateral-sold),
+          total-debt-burned: (+ (get total-debt-burned auction) usda-to-use)
+        }))
+
+        true
+      )
+    )
+
+    ;; Try to finalize
+    (try! (finalize-liquidation auction-id vault-manager coll-type))
+
+    (print { type: "auction", action: "burn-usda", data: auction })
+    (ok usda-to-use)
+  )
+)
+
+;; @desc finalize liquidation so rest of collateral becomes available to vault owner
+;; @param auction-id; the auction
+;; @param vault-manager; vault manager
+;; @param coll-type; collateral types
+(define-public (finalize-liquidation
+  (auction-id uint)
+  (vault-manager <vault-manager-trait>)
+  (coll-type <collateral-types-trait>)
+)
+  (let (
+    (auction (get-auction-by-id auction-id))
+  )
+    (asserts! (not (shutdown-activated)) (err ERR-EMERGENCY-SHUTDOWN-ACTIVATED))
+    (asserts! (is-eq (contract-of vault-manager) (unwrap-panic (contract-call? .arkadiko-dao get-qualified-name-by-name "freddie"))) (err ERR-NOT-AUTHORIZED))
+    (asserts! (is-eq (contract-of coll-type) (unwrap-panic (contract-call? .arkadiko-dao get-qualified-name-by-name "collateral-types"))) (err ERR-NOT-AUTHORIZED))
+
+    ;; Sell off DIKO if needed
+    (try! (sell-diko auction-id))
```

Comment:
REQ: I think there's something odd going on here.

Imagine the following scenario:
1. A vault gets liquidated but there is not enough USDA to buy all collateral in the fund
2. Someone calls this immediately after. Since there is not enough USDA, it mints DIKO and sells that off

This is a bit odd since we should only mint and sell DIKO if the total collateral amount in the vault is not sufficient to pay back the debt, and not whenever there is not sufficient USDA. In other words, DIKO minting should be considered in isolation, per vault. Imagine a vault of 1M STX needs to get liquidated in current model, this would completely destroy the DIKO price.

Reply:
On L321 in function `sell-diko` we check if all collateral is sold. Only if that's the case, DIKO will be sold.

Reply:
Oh nice, I missed that. Should we do something like >= (or <=) instead of `is-eq`? Not sure if something can go wrong but it seems a bit more loose and less error prone